### PR TITLE
Automatic differentiation for Wasserstein distance

### PR DESCRIPTION
**Pull request by mglisse**
Maybe some sinkhorn-based version (there are some available on the net) would be smoother and work better for optimization, but since this is simple, it seems worth providing.
An idea would be to split the test into 2 files, so that missing torch/eagerpy does not disable the whole test. It didn't seem worth the trouble, but it could be relevant say for the debian package where they may get a POT package soon but seem unlikely to get torch.

### CODE REVIEW

File: src/python/gudhi/wasserstein/wasserstein.py [hunk 1]
```diff
@@ -64,17 +64,31 @@ def _build_dist_matrix(X, Y, order, internal_p):
...
     :param order: exponent for Wasserstein. Default value is 2.
     :param internal_p: Ground metric on the (upper-half) plane (i.e. norm L^p in R^2); Default value is 2 (Euclidean norm).
+    :param enable_autodiff: If X is torch.tensor, tensorflow.Tensor or jax.numpy.ndarray, make the computation
+        transparent to automatic differentiation.
+    :type enable_autodiff: bool
     :returns: float, the total persistence of the diagram (that is, its distance to the empty diagram).
     '''
-    return np.linalg.norm(_dist_to_diag(X, internal_p), ord=order)
+    if enable_autodiff:
+        import eagerpy as ep
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/wasserstein/wasserstein.py:84]: I think raising an exception is a better behavior than printing a message and returning some wrong value (the situation may be a bit different for a function that shows a graphics window and whose return objet is often ignored). ImportError also comes with a pretty clear message, it doesn't seem worth expanding on it.
On the other hand, I really should document this dependency on eagerpy. While running the code will show it, the user should have been informed earlier by the documentation.
(actually, I was thinking of bundling a fork of eagerpy with gudhi, because it is small, pure python, and I may need to patch it in places, but that's for later)

File: src/python/gudhi/wasserstein/wasserstein.py [hunk 2]
```diff
@@ -93,30 +110,39 @@ def wasserstein_distance(X, Y, matching=False, order=2., internal_p=2.):
...
+        X = X_orig.numpy()
+        Y = Y_orig.numpy()
     M = _build_dist_matrix(X, Y, order=order, internal_p=internal_p)
     a = np.ones(n+1) # weight vector of the input diagram. Uniform here.
     a[-1] = m
     b = np.ones(m+1) # weight vector of the input diagram. Uniform here.
     b[-1] = n
 
     if matching:
+        assert not enable_autodiff, "matching and enable_autodiff are currently incompatible"
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/wasserstein/wasserstein.py:145]: Add more code to handle that case? It isn't hard, just didn't seem like a priority.

File: src/python/test/test_wasserstein_distance.py [hunk 3]
```diff
@@ -80,14 +80,33 @@ def approx(x):
...
+
+    diag1 = torch.tensor([[2.7, 3.7], [9.6, 14.0], [34.2, 34.974]], requires_grad=True)
+    diag2 = torch.tensor([[2.8, 4.45], [9.5, 14.1]], requires_grad=True)
+    diag3 = torch.tensor([[2.8, 4.45], [9.5, 14.1]], requires_grad=True)
+    assert diag1.grad is None and diag2.grad is None and diag3.grad is None
+    dist1 = pot(diag1, diag2, internal_p=2, order=2, enable_autodiff=True)
+    dist2 = pot(diag3, torch.tensor([]), internal_p=2, order=2, enable_autodiff=True)
+    dist1.backward()
+    dist2.backward()
+    assert not torch.isnan(diag1.grad).any() and not torch.isnan(diag2.grad).any() and not torch.isnan(diag3.grad).any()
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/test/test_wasserstein_distance.py:112]: First it checks that grad actually exists, that's the main element. As a bonus it checks that is doesn't have any NaN (I should add a case where the distance is 0 to make that check interesting, but then it may require a patched version of eagerpy...). Adding an explicit simple example makes sense, yes.

File: src/python/gudhi/wasserstein/wasserstein.py [hunk 4]
```diff
@@ -85,6 +99,9 @@ def wasserstein_distance(X, Y, matching=False, order=2., internal_p=2.):
     :param order: exponent for Wasserstein; Default value is 2.
     :param internal_p: Ground metric on the (upper-half) plane (i.e. norm L^p in R^2);
                        Default value is 2 (Euclidean norm).
+    :param enable_autodiff: If X and Y are torch.tensor, tensorflow.Tensor or jax.numpy.ndarray, make the computation
+        transparent to automatic differentiation.
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/wasserstein/wasserstein.py]: > you cannot "create" new points

Better not, or the gradient (direction which increases the function) might tell me to pull an infinite number of points from the diagonal.

> nor merge a point into the diagonal

When you compute the gradient, you only look at what happens in an arbitrarily small neighborhood, there is no such thing as reaching some position over there. But following the (opposite of the) gradient can push a point in the direction of the diagonal, and in a gradient descent, it can end up on the diagonal.

**Comment by mglisse** [comment]
> There should be an `Optional: Eagerpy`, or `Eagerpy for enable_autodiff` in the `requires` field in file `src/python/doc/wasserstein_distance_sum.inc` like we did for [Witness compllex](https://gudhi.inria.fr/python/latest/#witness-complex) for instance.

Hmm, we didn't do it for point_cloud. How should it look like there, different functions (and different parameters of functions) have different dependencies, in the list CGAL+Eigen, numpy, scipy, sklearn, hnswlib, pykeops+torch, eagerpy (maybe I forgot one). And some of the dependencies are build-time only while others are runtime only. Let's discuss that in an issue.
